signer/src/dkg/verification.rs

```diff
@@ -0,0 +1,804 @@
+{
+    /// Creates a new [`StateMachine`] with the given [`FrostCoordinator`],
+    /// aggregate key, and timeout.
+    pub fn new<X>(coordinator: FrostCoordinator, aggregate_key: X, timeout: Duration) -> Self
+    where
+        X: Into<XOnlyPublicKey>,
+    {
+        let aggregate_key = aggregate_key.into();
+
+        Self {
+            aggregate_key,
+            coordinator,
+            wsts_messages: HashMap::new(),
+            created_at: Instant::now(),
+            timeout,
+            state: State::Idle,
+            _error: std::marker::PhantomData,
+        }
+    }
+
+    /// Processes a WSTS message, updating the internal state of the
+    /// [`FrostCoordinator`].
+    ///
+    /// - Will return an error if an invalid operation is attempted given the
+    ///   current state of the [`StateMachine`].
+    /// - Poll [`Self::state`] after a successfull call to this function to get
+    ///   the current state of the [`StateMachine`], which will be updated as
+    ///   messages are processed.
+    /// - If the instance is in an end-state, an error will be returned.
+    /// - Upon success, the resulting signature can be retrieved via
+    ///   [`Self::state`].
+    /// - Will process all eligible pending messages given the current state of
+    ///   the [`FrostCoordinator`].
+    pub fn process_message<M>(&mut self, sender: PublicKey, msg: M) -> Result<(), TError>
+    where
+        M: Into<wsts::net::Message> + std::fmt::Debug,
+    {
+        // Set our state to expired if we've passed the timeout.
+        if self.is_expired() {
+            self.state = State::Expired;
+        }
+
+        // Assert that we're in a valid state to process messages.
+        self.assert_valid_state()?;
+
+        // Enqueue the message to be processed. If the constraints for
+        // enqueueing the message are not met, an error will be returned and we
+        // don't make any further updates or process anything.
+        self.enqueue_message(sender, msg.into())?;
+
+        // We've enqueued at least 1 message, so just make sure we're in the
+        // `Signing` state.
+        self.state = State::Signing;
+
+        // Iterate through all of the queued messages and process the relevant
+        // ones. This loop will first process all messages applicable for the
+        // current state, and then continue to process messages until there are
+        // no more state changes. This is necessary because the state machine
+        // can receive out-of-order messages, and we need to ensure that we
+        // process all messages that we can before returning, including handling
+        // state transitions.
+        loop {
+            let current_processable_message_type = self.current_processable_message_type();
+            self.process_queued_messages()?;
+            if current_processable_message_type == self.current_processable_message_type() {
+                break;
+            }
+        }
```

Comment:
if `self.process_queued_messages()` were to return the number of processed messages (or even if we did process at least one message) we could simplify this by iterating until we stop doing things

Reply:
Yeah or just a bool, even easier, good catch 👍

Reply:
[c9c66c4](https://github.com/stacks-network/sbtc/pull/1324/commits/c9c66c4c683269d36267fffc020ac3ed7959ab81)